xds/src/main/java/io/grpc/xds/internal/sds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.sds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.Iterables;
+import com.google.protobuf.UInt32Value;
+import io.grpc.Attributes;
+import io.grpc.netty.GrpcHttp2ConnectionHandler;
+import io.grpc.netty.InternalProtocolNegotiationEvent;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.ProtocolNegotiationEvent;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.ConnectionSourceType;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+
+/**
+ * Handles L4 filter chain match for the connection based on the xds configuration.
+ * */
+public final class FilterChainMatchingHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger log = Logger.getLogger(FilterChainMatchingHandler.class.getName());
+  private final GrpcHttp2ConnectionHandler grpcHandler;
+  private final FilterChainSelector selector;
+  private final ProtocolNegotiator delegate;
+
+  public static final Attributes.Key<SslContextProviderSupplier>
+          ATTR_SERVER_SSL_CONTEXT_PROVIDER_SUPPLIER =
```

Comment:
Similar to my other comment, should this be in `InternalXdsAttributes` ?

Reply:
hmm, would it better to be moved to sds internal package? I would soon do so. Because
On one hand, you seems wanted to do similar thing for the client:
https://github.com/grpc/grpc-java/blob/d4fa0ecc07495097453b0a2848765f076b9e714c/xds/src/main/java/io/grpc/xds/InternalXdsAttributes.java#L34

on the other, ejona@ seemed to intuitively like to put it there when we discussed package locations for filter chain matching handler:
```
It would be okay if we used a separate attribute for the security provider and had the sds code define that attribute.
```